File: lib/recording/gcs-path.js
/**
 * GCS path helpers shared by every agent that uploads recordings and by the
 * REST download endpoint. Keeping these here means there is exactly one
 * formula for the bucket / object name a recording lands at.
 */

/**
 * @typedef {Object} GcsPath
 * @property {string} bucket
 * @property {string} prefix Always empty or ending with a single trailing slash.
 */

/**
 * Parse a `gs://bucket[/prefix]` URL into `{ bucket, prefix }`.
 *
 * @param {string} baseUrl
 * @returns {GcsPath}
 */
export function parseGcsPath(baseUrl) {
  if (!baseUrl.startsWith('gs://')) {
    throw new Error('Recording storage path must be a gs:// URL');
  }

  const withoutScheme = baseUrl.slice('gs://'.length);
  const firstSlash = withoutScheme.indexOf('/');
  if (firstSlash === -1) {
    return { bucket: withoutScheme, prefix: '' };
  }

  const bucket = withoutScheme.slice(0, firstSlash);
  let prefix = withoutScheme.slice(firstSlash + 1);
  if (prefix.length > 0 && !prefix.endsWith('/')) {
    prefix += '/';
  }
  return { bucket, prefix };
}

/**
 * Default base URL used when `RECORDING_STORAGE_PATH` is unset. Mirrors the
 * historical LiveKit default so a fresh deploy without explicit configuration
 * keeps working.
 *
 * @returns {string}
 */
export function defaultRecordingBaseUrl() {
  const { RECORDING_STORAGE_PATH, NODE_ENV } = process.env;
  return (
    RECORDING_STORAGE_PATH || `gs://llm-voice/${NODE_ENV || 'development'}-recordings`
  );
}

/**
 * Build the GCS object name for a given call. Recordings are always OGG.
 *
 * @param {string} prefix Empty string or trailing-slash prefix from {@link parseGcsPath}.
 * @param {string} callId
 * @returns {string}
 */
export function objectNameFor(prefix, callId) {
  return `${prefix}${callId}.ogg`;
}

File: agents/pipecat/pipecat_aplisay/recording/gcs_path.py
"""GCS path helpers — sibling of ``lib/recording/gcs-path.js``."""

from __future__ import annotations

import os


def parse_gcs_path(base_url: str) -> tuple[str, str]:
    """Split a ``gs://bucket[/prefix]`` URL into ``(bucket, prefix)``.

    ``prefix`` is either empty or ends with a single ``/`` so callers can
    string-concat the object name without further normalisation.
    """
    if not base_url.startswith("gs://"):
        raise ValueError("Recording storage path must be a gs:// URL")

    without_scheme = base_url[len("gs://") :]
    first_slash = without_scheme.find("/")
    if first_slash == -1:
        return without_scheme, ""

    bucket = without_scheme[:first_slash]
    prefix = without_scheme[first_slash + 1 :]
    if prefix and not prefix.endswith("/"):
        prefix += "/"
    return bucket, prefix


def default_recording_base_url() -> str:
    """Default GCS base URL when ``RECORDING_STORAGE_PATH`` is unset.

    Mirrors the JS default exactly so deployments without explicit
    configuration land on the same bucket regardless of which agent wrote
    the file.
    """
    storage_path = os.environ.get("RECORDING_STORAGE_PATH")
    if storage_path:
        return storage_path
    node_env = os.environ.get("NODE_ENV", "development")
    return f"gs://llm-voice/{node_env}-recordings"


def object_name_for(prefix: str, call_id: str) -> str:
    """Build the GCS object name for a recording. Always ``.ogg``."""
    return f"{prefix}{call_id}.ogg"
